src/renderer/components/settings-execution.tsx

```diff
@@ -59,36 +101,201 @@ export class ExecutionSettings extends React.Component<ExecutionSettingsProps> {
+      appState.environmentVariables = vars.filter((v) => v !== '');
+    }
   }
 
   /**
-   * Handles a change in the environment variables run with the Electron executable.
+   * Handles a change in the execution flags or environment variables
+   * run with the Electron executable.
    *
    * @param {React.ChangeEvent<HTMLInputElement>} event
+   * @param {SettingItemType} type
    */
-  public handleEnvironmentVariableChange(
-    event: React.FormEvent<HTMLInputElement>,
+  public handleSettingsItemChange(
+    event: React.ChangeEvent<HTMLInputElement>,
+    type: SettingItemType,
   ) {
+    const { name, value } = event.currentTarget;
+
+    if (type === SettingItemType.Flag) {
+      this.setState((prevState) => ({
+        executionFlags: {
+          ...prevState.executionFlags,
+          [name]: value,
+        },
+      }));
+    } else {
+      this.setState((prevState) => ({
+        environmentVariables: {
+          ...prevState.environmentVariables,
+          [name]: value,
+        },
+      }));
+    }
+  }
+
+  /**
+   * Adds a new settings item input field.
+   *
+   * @param {SettingItemType} type
+   */
+  private addNewSettingsItem(type: SettingItemType) {
+    if (type === SettingItemType.Flag) {
+      const flagsArray = Object.entries(this.state.executionFlags);
+
+      this.setState((prevState) => ({
+        executionFlags: {
+          ...prevState.executionFlags,
+          [`flag-${flagsArray.length}`]: '',
+        },
+      }));
+    } else {
+      const varsArray = Object.entries(this.state.environmentVariables);
+
+      this.setState((prevState) => ({
+        environmentVariables: {
+          ...prevState.environmentVariables,
+          [`var-${varsArray.length}`]: '',
+        },
+      }));
+    }
+  }
+
+  /**
+   * Handle a change to the package manager used to install modules when running
+   * Fiddles;
+   *
+   * @param {React.FormEvent<HTMLInputElement>} event
+   */
+  private handlePMChange = (event: React.FormEvent<HTMLInputElement>) => {
+    const { appState } = this.props;
     const { value } = event.currentTarget;
 
-    const variables = value.split('|');
-    this.props.appState.environmentVariables = variables;
+    appState.packageManager = value as IPackageManager;
+  };
+
+  public renderDeleteItem(idx: string, type: SettingItemType): JSX.Element {
+    const clickFn = () => {
+      if (type === SettingItemType.Flag) {
+        const flags = this.state.executionFlags;
+        if (Object.keys(flags).length === 1) {
+          flags[idx] = '';
+        } else {
+          delete flags[idx];
+        }
+
+        this.setState({
+          executionFlags: flags,
+        });
+      } else {
+        const vars = this.state.environmentVariables;
+        if (Object.keys(vars).length === 1) {
+          vars[idx] = '';
+        } else {
+          delete vars[idx];
+        }
+
+        this.setState({
+          environmentVariables: vars,
```

Comment:
It _looks_ like we're calling setState() with the same variable that's already in state (ie this.state.environmentVariables). Is this call to setState() redundant?

Reply:
hmm - not sure what you mean, since we're modifying that object directly above this 🤔

Reply:
Right we are, but when we modify `this.state.environmentVariable` isn't that an `@observable` such that the component already knows about the modifications we made even without calling setState()?

Reply:
@ckerr no - this class-level state is not an observable like that on `appState` :)